### DISCUSSION TITLE
Easy way to get started to contribute ?

### DISCUSSION THREAD

**Discussion by DzzD** [supermerill/SuperSlicer#1873] [Community contribution]
Hello,

I would like to contribute to Super Slicer, for some ressources but also as a developer, unfortunatly I cant find any clear path to get started and I am confused in the relation betwenn Super Slicer/Slic3r/Prusa Slicer.

Does anyone can give me a clear path to get started (more for window, but I am also interrested by general information )?

I have read the windows article on this reository but I would like more or clearer exlanatinon plz....

**Comment by supermerill** [comment]
Hello
First, you have to succeed to compile it.
Then, you can modify a little thing you want to work on, like a new infill pattern.
after that, ask for a pull request to have your contribution merged.

I don't know what could motivate you, i can help you by explaining how some part of the software works, if you have an idea to implement.

**Comment by DzzD** [comment]
I decided tonight to give a try to compile, but failed on cmake :/

![image](https://user-images.githubusercontent.com/12198339/142043603-54bc70bb-3b5d-4518-b26c-80602ed8cb8b.png)

I use this line : cmake .. -G "Visual Studio 17 2022" -A x64  

because the 16 2019 was giving me another error

![image](https://user-images.githubusercontent.com/12198339/142043843-7dcd2656-f77d-45ac-b00b-2ab555af94a6.png)

Can you help ?


EDIT: forget aboout this I will install 2019 tool

**Reply by supermerill** [reply]
you have to install visual studio 2019

**Reply by DzzD** [reply]
I was thinking that 2022 will be able to compile for 2019 :/ so I need to find a place where to find it

**Reply by supermerill** [reply]
I didn't try VS2022, so maybe, maybe not.If there is some problems, first thing is to stick with something that is known to work.

**Reply by DzzD** [reply]
fo sure it can compile "Visual Studio 17 2022" , I must missing some component as it dont find "Visual Studio 16 2019"

**Comment by DzzD** [comment]
So I manage to execute cmake fine with 2022, but I got 7 error in the next step **msbuild process**, here is one : 


       "C:\Dev\SuperSlicer\deps\build\ALL_BUILD.vcxproj" (cible par défaut) (1) ->
       "C:\Dev\SuperSlicer\deps\build\deps.vcxproj" (cible par défaut) (3) ->
       "C:\Dev\SuperSlicer\deps\build\dep_cereal.vcxproj" (cible par défaut) (14) ->
         C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(
       242,5): error MSB8066: la build personnalisée de 'C:\Dev\SuperSlicer\deps\build\CMakeFiles\68d68a99c5c535af854c5
       0d82b97a5f8\dep_cereal-mkdir.rule;C:\Dev\SuperSlicer\deps\build\CMakeFiles\68d68a99c5c535af854c50d82b97a5f8\dep_
       cereal-download.rule;C:\Dev\SuperSlicer\deps\build\CMakeFiles\68d68a99c5c535af854c50d82b97a5f8\dep_cereal-update
       .rule;C:\Dev\SuperSlicer\deps\build\CMakeFiles\68d68a99c5c535af854c50d82b97a5f8\dep_cereal-patch.rule;C:\Dev\Sup
       erSlicer\deps\build\CMakeFiles\68d68a99c5c535af854c50d82b97a5f8\dep_cereal-configure.rule;C:\Dev\SuperSlicer\dep
       s\build\CMakeFiles\68d68a99c5c535af854c50d82b97a5f8\dep_cereal-build.rule;C:\Dev\SuperSlicer\deps\build\CMakeFil
       es\68d68a99c5c535af854c50d82b97a5f8\dep_cereal-install.rule;C:\Dev\SuperSlicer\deps\build\CMakeFiles\efd2c0bbac0
       b3e326198571208a7f310\dep_cereal-complete.rule;C:\Dev\SuperSlicer\deps\build\CMakeFiles\c35c6af29a8fb793aed2fba8
       d6617004\dep_cereal.rule;C:\Dev\SuperSlicer\deps\CMakeLists.txt' s'est arrêtée. Code 1. [C:\Dev\SuperSlicer\deps
       \build\dep_cereal.vcxproj]

any idea ?

**Reply by supermerill** [reply]
no, never happened to me.

**Reply by DzzD** [reply]
I still got one error.... when building dependency : 

```
ÉCHEC de la build.

       "C:\Dev\SuperSlicer\deps\build\ALL_BUILD.vcxproj" (cible par défaut) (1) ->
       "C:\Dev\SuperSlicer\deps\build\deps.vcxproj" (cible par défaut) (3) ->
       "C:\Dev\SuperSlicer\deps\build\dep_openvdb.vcxproj" (cible par défaut) (17) ->
       "C:\Dev\SuperSlicer\deps\build\dep_blosc.vcxproj" (cible par défaut) (19) ->
       (CustomBuild cible) ->
         CUSTOMBUILD : warning : squelched 54 whitespace errors [C:\Dev\SuperSlicer\deps\build\dep_blosc.vcxproj]
         CUSTOMBUILD : warning : 4 lines applied after fixing whitespace errors. [C:\Dev\SuperSlicer\deps\build\dep_blo
       sc.vcxproj]


       "C:\Dev\SuperSlicer\deps\build\ALL_BUILD.vcxproj" (cible par défaut) (1) ->
       "C:\Dev\SuperSlicer\deps\build\deps.vcxproj" (cible par défaut) (3) ->
       "C:\Dev\SuperSlicer\deps\build\dep_CGAL.vcxproj" (cible par défaut) (4) ->
       "C:\Dev\SuperSlicer\deps\build\dep_boost.vcxproj" (cible par défaut) (9) ->
       (CustomBuild cible) ->
         C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(
       242,5): error MSB8066: la build personnalisée de 'C:\Dev\SuperSlicer\deps\build\CMakeFiles\5119d4c7bcf2580fb03bc
       6647c79bc61\dep_boost-configure.rule;C:\Dev\SuperSlicer\deps\build\CMakeFiles\5119d4c7bcf2580fb03bc6647c79bc61\d
       ep_boost-build.rule;C:\Dev\SuperSlicer\deps\build\CMakeFiles\5119d4c7bcf2580fb03bc6647c79bc61\dep_boost-install.
       rule;C:\Dev\SuperSlicer\deps\build\CMakeFiles\efd2c0bbac0b3e326198571208a7f310\dep_boost-complete.rule;C:\Dev\Su
       perSlicer\deps\build\CMakeFiles\c35c6af29a8fb793aed2fba8d6617004\dep_boost.rule' s'est arrêtée. Code 1. [C:\Dev\
       SuperSlicer\deps\build\dep_boost.vcxproj]

    2 Avertissement(s)
    1 Erreur(s)
```

It seems related to bootstrap : 
"Please consult bootstrap.log for further diagnostics."

wich contain : 
```
...
...
property-set.c
sequence.c
order.c
G‚n‚ration de code en cours...
###
### Unknown toolset: vcunk
###
### Known toolsets are: acc, borland, cc, como, clang, darwin, gcc, gcc-nocygwin, intel-darwin, intel-linux, intel-win32, kcc, kylix, metrowerks, mipspro, msvc, qcc, pathscale, pgi, sun, sunpro, tru64cxx, vacpp, xlcpp, vc7, vc8, vc9, vc10, vc11, vc12, vc14, vc141, vc142, vmsdecc
###
###
### Unknown toolset: vcunk
###
### Known toolsets are: acc, borland, cc, como, clang, darwin, gcc, gcc-nocygwin, intel-darwin, intel-linux, intel-win32, kcc, kylix, metrowerks, mipspro, msvc, qcc, pathscale, pgi, sun, sunpro, tru64cxx, vacpp, xlcpp, vc7, vc8, vc9, vc10, vc11, vc12, vc14, vc141, vc142, vmsdecc
###
```

**Reply by supermerill** [reply]
seems like it's not compatible with 2022

**Comment by DzzD** [comment]
Ok so I failed to compile dependencies with VS 2022 .. for only 2 modules :(

If anyone is searching for the old download of Visual Studio 2019, it can be downloaded here : https://docs.microsoft.com/fr-fr/visualstudio/releases/2019/history

**Comment by DzzD** [comment]
So now that I can compile & run ! is there some kind of documentation somewhere that you can point to me ?

What are each package for ?

![image](https://user-images.githubusercontent.com/12198339/142140963-7ab231d0-6669-444b-bf15-7a744da1d50c.png)

Can you show me a little example so it will be easier to sart ? like where package/file can I add a text field to the printer settings screen ?

**Reply by supermerill** [reply]
the gui stuff is under slic3r/GUI
the core stuff that can work without a gui is under libslic3r
the main method for slicing is Print::process, and that dispatch the calls to all objects, step by step.
The gcode creation is currently done by Gcode,  but i'm currently moving stuff from here.

If you want something simple:
 - To add a pattern, you have to create your pattern class in libslic3r/Fill
 - dont forget to add your new file to the cmakelist.
 - makes it extends the Fill abstract class. (check FillGyroid)
 - you can extends the _fill_surface_single method if you need something simple: you just have to add some series of lines, and they will be transformed to extrusions automatically. (check FillGyroid)
 - If you need something complex with multiple process, you can extends fill_surface_extrusion (check FillSmooth)

All of these are called by Fill.cpp
The Fill class and the FillParams are in FillBAse.h / cpp , don't ask why, i don't know.


a Point is a two-element array of coord_t, which are a int64_t. To get a coord_t from a mm value, you have to use scale_t(mm_value). To have the mm value of a coord_t, you have to use unscaled(coord_val). Try to use a double when storing a mm value, coord_t when storing a scaled value, and coordf_t when storing a double representation of a scaled value (useful when storing a square distance)
a polyline is a serie of point
a polygon is a polyline but with a virtual segment between the last and first point.
a ExPolygon is a polygon with holes inside. Prefer using that, as it's safer than polygons with different directions.
clipper is used to do stuff with polygons (and polylines), like offset, diff, intersection and union. See PeriemterGenerator for exemples.

**Reply by DzzD** [reply]
Thank's that helps !

**Reply by supermerill** [reply]
An random idea of a thing to implement as a pattern:
- search a motif, like https://www.istockphoto.com/fr/vectoriel/lignes-g%C3%A9om%C3%A9triques-abstrait-motif-transparent-avec-motif-antique-grecque-gm962186356-262777674
- create a method that creates the pattern (like for gyroid) on a square.
- clip it to your surface, like for gyroid.
- add links if needed.
- test it!

**Reply by supermerill** [reply]
Oh, I forgot to explain how the settings works.
To create an option, 
* You have to create it into PrintConfig.cpp (ordered by name), copy a setting of the same type and modify the names.
* You then have to add it to printconfig.hpp, in the correct class:
   * printConfig is available evrywhere, but can't be modified in a modifier
   * Objectconfig is only avaialable when you have an object, and can be modified in object-modifier
   * PrintRegionConfig is only available in a region, and can be modified in all sort of modifier (shape, height)
   * GcodeConfig is for most of filament & printer settings. Filament and printer setting aren't available in modifiers anyway.
 * You have to add the name into the correct Preset section (print, printer, filament, slaprint, slamaterial, slaprinter)
 * Then add a new setting line in the corresponding resources/ui_layout/[tab_name].ui file.
 
Note that enum object also need the enum spec in printconfig.hpp, and many new codepath in many files. search for all occurence of "DenseInfillAlgo"  or any other enum to see where to add stuff to makes it work.
If you want to just add a new option to the enum setting, add the new enum name and the new stuff line into the get_enum_values() in PrintConfig.hpp, also add a new enum_values line and enum_labels line in PrintConfig.cpp.

For accessing a setting, if should have access to the object containing the setting to be able to call config().my_setting.
Sometimes, you have only access to a DynamicConfig, which is a container of various settings. Check what it contains in debug mode.
FullPrintConfig contains all ettings for a given region in a  given object. With that, you can call get_computed_value(field_name, extruder_id) to get the value of a complicated field with many % dependencies.

**Reply by DzzD** [reply]
Wow, PrintConfig.cpp is really scaring :) 

And about the GUI, where do I add a setting on a screen for example how do I place my new setting on the upper right corner of the Print Setting ?

Sorry for all those noob question, but I havn't done C++ for maybe last 20 years, I used tens of others languages but no c++ since 2001/2002 :)

**Reply by supermerill** [reply]
you modify the resources/ui_layout/XXX.ui
forgot about that.
doc about these files is on the readme.md

**Reply by DzzD** [reply]
perfect I suppose there is no visual editor ?

**Reply by supermerill** [reply]
no, just restart.

**Reply by DzzD** [reply]
And for here : 
![image](https://user-images.githubusercontent.com/12198339/142253330-39b05861-b6ae-4bab-b3ac-a517061ac569.png)

seems that the ui files begin inside the panel

**Reply by supermerill** [reply]
That is much more difficult thing to do. You have to modify the Tab class. You're on your own with that.

**Reply by DzzD** [reply]
ok, thanks again as I found exacly where it is in Tab.cpp, now the problem is to re-learn c++ and it should be ok ;)

**Comment by DzzD** [comment]
This is going well, I got the GUI mostly done, now I need to play whith "Preset" and it, seems that it will be the hard part with all the hierarchical strcutre and so.

**Reply by supermerill** [reply]
You can "just" "apply" your modification to the current full print config. The hard part is to re-apply it when needed and to be able to make the gui to not be tagged as "modified".

**Comment by DzzD** [comment]
Hello, How are we supposed to send you a patch ?

**Reply by supermerill** [reply]
you can open a pull request

**Reply by DzzD** [reply]
I know but this is for a tiny change unrelated from what I am working on (sub/version preset), this is just two little patches, I was wondering if there was a way to send it, but I can create a pull request if required

**Reply by tome9111991** [reply]
I will find the best solution is you fork SuSi and make change on your forked repo.
If you done you can create a Pull request.
Merill can see the changes and cand merge with one click.

**Reply by DzzD** [reply]
Yes I already did it and I am working on it, I will do a pull request but I was just wondering if sometime little patch could be applied without all the complexity of a fork etc..., (note that I am also learning at the same time for one of my own project that I open sourced this weeks ! https://github.com/DzzD/TiGL ;) )

**Reply by supermerill** [reply]
Your local repo is a fork.
You can submit any branch as a pullrequest. If you add some commit in a branch (moving the branch tag), then the pull request will update accordingly.
You have a little thing to submit? commt, create a branch here, push and click on "submit pr".
The little commit is in the middle of a a big work-in-progress? create a branch at my master, then cherry-pick the commit, push the branch, click on "submit pr"
With tortoisegit, it's one click per step. Done in ~20 sec.
Giving a patch to incorporate nececitate more work on your side & my side.

**Reply by DzzD** [reply]
ok, thanks for this summary
